src/connection.c

```diff
@@ -160,12 +163,99 @@ static void handleConnected(const redisAsyncContext *c, int status)
         return;
     }
 
-    /* Call explicit connect callback (even if failed) */
+    /* Call callback even when failed */
     if (conn->connect_callback) {
         conn->connect_callback(conn);
     }
 }
 
+static void handleAuth(redisAsyncContext *c, void *r, void *privdata)
+{
+    Connection *conn = privdata;
+
+    redisReply *reply = r;
+    if (!reply) {
+        LOG_ERROR("Redis connection authentication failed: connection died");
+        ConnMarkDisconnected(conn);
+    } else if (reply->type == REDIS_REPLY_ERROR) {
+        LOG_ERROR("Redis connection authentication failed: %s", reply->str);
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    } else if (reply->type != REDIS_REPLY_STATUS || strcmp("OK", reply->str) != 0) {
+        LOG_ERROR("Redis connection authentication failed: unexpected response");
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    } else {
+        return connectionSuccess(conn);
+    }
+
+    return connectionFailure(conn);
+}
+
+/* Connection with authentication
+ *
+ * If the connection status is ok, will attempt to authenticate and only then call the connection callback
+ * If connection status is not ok, will immediately transition connection to CONN_CONNECT_ERROR
+ * User callback is called in all cases except when the connection has already been terminated externally
+ */
+
+static void handleConnectedWithAuth(Connection *conn, int status)
+{
+    if (status == REDIS_OK) {
+        /* don't try to authenticate if terminating */
+        if (conn->flags & CONN_TERMINATING) {
+            return connectionSuccess(conn);
```

Comment:
Why we are calling connectionSuccess(); here? Shouldn't we just return?

Reply:
to set the flags, as if we didn't do auth.  not sure it matters.

Reply:
I think calling connectionSuccess() here gives wrong impression as if like it actually connected. Probably, we should just return. 

We still need to fix the issue with CONN_TERMINATING, maybe as part of another PR.

Reply:
in looking at the code, it seems the fix is that ConnIsIdle has to be expanded for checking the terminating flag.

Reply:
@sjpotter I think we also need to consider the visibility of the connection state. For example, differentiate "connecting" from "authenticating" and "connected".

This also makes me realize we should probably track auth errors.

Reply:
I think I addressed it all in the latest commit

Reply:
ok, didn't see yossi's last comment before I updated, will look into addressing that too.